File: 2nd Year/Term 2/Data Structures & Algorithms/MS1/LinkedList/Recursion/LinkedListRecursion.py
# Recursion in Linked List - Recursive search by brand (MotorPH)


class Node:
    def __init__(self, brand, model, quantity):
        self.data = {"brand": brand, "model": model, "quantity": quantity}
        self.next = None


def recursive_search(node, brand):
    """Search linked list by brand using recursion."""
    if node is None:
        return None
    if node.data["brand"] == brand:
        return node.data
    return recursive_search(node.next, brand)


def display(head):
    """Display linked list from head."""
    if head is None:
        print("  (empty)")
        return
    current = head
    while current:
        d = current.data
        print(f"  - {d['brand']} {d['model']}: {d['quantity']} units")
        current = current.next


def main():
    # Build linked list: Honda -> Yamaha -> Suzuki
    head = Node("Honda", "Click 125", 10)
    head.next = Node("Yamaha", "NMAX 155", 5)
    head.next.next = Node("Suzuki", "Burgman Street", 7)

    print("Linked list:")
    display(head)

    # Recursive search by brand
    print("\nRecursive search for 'Yamaha':")
    result = recursive_search(head, "Yamaha")
    print(f"  {result if result else 'Not found'}")

    print("\nRecursive search for 'Honda':")
    result = recursive_search(head, "Honda")
    print(f"  {result if result else 'Not found'}")

    print("\nRecursive search for 'Kawasaki':")
    result = recursive_search(head, "Kawasaki")
    print(f"  {result if result else 'Not found'}")


if __name__ == "__main__":
    main()

File: 2nd Year/Term 2/Data Structures & Algorithms/MS1/Old files (mistake)/LinkedList/Recursion/LinkedListRecursion.java
package LinkedList.Recursion;

/**
 * Recursion in Linked List - Recursive search by brand (MotorPH)
 */
public class LinkedListRecursion {

    static class Node {
        InventoryItem data;
        Node next;

        Node(String brand, String model, int quantity) {
            this.data = new InventoryItem(brand, model, quantity);
            this.next = null;
        }
    }

    record InventoryItem(String brand, String model, int quantity) {}

    /** Search linked list by brand using recursion. */
    static InventoryItem recursiveSearch(Node node, String brand) {
        if (node == null) {
            return null;
        }
        if (node.data.brand().equals(brand)) {
            return node.data;
        }
        return recursiveSearch(node.next, brand);
    }

    /** Display linked list from head. */
    static void display(Node head) {
        if (head == null) {
            System.out.println("  (empty)");
            return;
        }
        Node current = head;
        while (current != null) {
            InventoryItem d = current.data;
            System.out.println("  - " + d.brand() + " " + d.model() + ": " + d.quantity() + " units");
            current = current.next;
        }
    }

    public static void main(String[] args) {
        // Build linked list: Honda -> Yamaha -> Suzuki
        Node head = new Node("Honda", "Click 125", 10);
        head.next = new Node("Yamaha", "NMAX 155", 5);
        head.next.next = new Node("Suzuki", "Burgman Street", 7);

        System.out.println("Linked list:");
        display(head);

        // Recursive search by brand
        System.out.println("\nRecursive search for 'Yamaha':");
        InventoryItem result = recursiveSearch(head, "Yamaha");
        System.out.println("  " + (result != null ? result : "Not found"));

        System.out.println("\nRecursive search for 'Honda':");
        result = recursiveSearch(head, "Honda");
        System.out.println("  " + (result != null ? result : "Not found"));

        System.out.println("\nRecursive search for 'Kawasaki':");
        result = recursiveSearch(head, "Kawasaki");
        System.out.println("  " + (result != null ? result : "Not found"));
    }
}
